tests/ops/op_math/test_controlled_op.py

```diff
@@ -0,0 +1,688 @@
+        assert op.basis == "Z"
+
+    @pytest.mark.parametrize(
+        "base, expected",
+        [
+            (qml.RX(1.23, wires=0), [(0.5, 1.0)]),
+            (qml.PhaseShift(-2.4, wires=0), [(1,)]),
+            (qml.IsingZZ(-9.87, (0, 1)), [(0.5, 1.0)]),
+        ],
+    )
+    def test_parameter_frequencies(self, base, expected):
+        """Test parameter-frequencies against expected values."""
+
+        op = Controlled(base, (3, 4))
+        assert op.parameter_frequencies == expected
+
+    def test_parameter_frequencies_no_generator_error(self):
+        """An error should be raised if the base doesn't have a generator."""
+        base = TempOperation(1.234, 1)
+        op = Controlled(base, 2)
+
+        with pytest.raises(
+            qml.operation.ParameterFrequenciesUndefinedError,
+            match=r"does not have parameter frequencies",
+        ):
+            op.parameter_frequencies
+
+    def test_parameter_frequencies_multiple_params_error(self):
+        """An error should be raised if the base has more than one parameter."""
+        base = TempOperation(1.23, 2.234, 1)
+        op = Controlled(base, (2, 3))
+
+        with pytest.raises(
+            qml.operation.ParameterFrequenciesUndefinedError,
+            match=r"does not have parameter frequencies",
+        ):
+            op.parameter_frequencies
+
+
+class TestQueuing:
+    """Test that Controlled operators queue and update base metadata."""
+
+    def test_queuing(self):
+        """Test that `Controlled` is queued upon initialization and updates base metadata."""
+        with qml.tape.QuantumTape() as tape:
+            base = qml.Rot(1.234, 2.345, 3.456, wires=2)
+            op = Controlled(base, (0, 1))
+
+        assert tape._queue[base]["owner"] is op
+        assert tape._queue[op]["owns"] is base
+        assert tape.operations == [op]
+
+    def test_queuing_base_defined_outside(self):
+        """Test that base isn't added to queue if its defined outside the recording context."""
+
+        base = qml.IsingXX(1.234, wires=(0, 1))
+        with qml.tape.QuantumTape() as tape:
+            op = Controlled(base, ("a", "b"))
+
+        assert len(tape._queue) == 1
+        assert tape._queue[op]["owns"] is base
+        assert tape.operations == [op]
+
+    def test_do_queue_false(self):
+        """Test that when `do_queue=False` is specified, the controlled op is not queued."""
+
+        base = qml.PauliX(0)
+        with qml.tape.QuantumTape() as tape:
+            op = Controlled(base, 1, do_queue=False)
+
+        assert len(tape._queue) == 0
+
+
+base_num_control_mats = [
+    (qml.PauliX("a"), 1, CNOT),
+    (qml.PauliZ("a"), 1, CZ),
+    (qml.SWAP(("a", "b")), 1, CSWAP),
+    (qml.PauliX("a"), 2, Toffoli),
+    (qml.RX(1.234, "b"), 1, CRotx(1.234)),
+    (qml.RY(-0.432, "a"), 1, CRoty(-0.432)),
+    (qml.RZ(6.78, "a"), 1, CRotz(6.78)),
+    (qml.Rot(1.234, -0.432, 9.0, "a"), 1, CRot3(1.234, -0.432, 9.0)),
+]
+
+
+class TestMatrix:
```

Comment:
Just a general comment, do you think it would be worth including tests where Controlled ops are placed in circuits (either in the measurement process or as part of the body of the circuit) and testing differentiability?

Reply:
This PR does not add device support, so these Controlled ops cannot be executed.  Hence the need for device support changes.